beacon-chain/rpc/statefetcher/fetcher.go

```diff
@@ -113,13 +114,29 @@ func (p *StateProvider) State(ctx context.Context, stateId []byte) (state.Beacon
 		}
 	case "finalized":
 		checkpoint := p.ChainInfoFetcher.FinalizedCheckpt()
-		s, err = p.StateGenService.StateByRoot(ctx, bytesutil.ToBytes32(checkpoint.Root))
+		startSlot, err := slots.EpochStart(checkpoint.Epoch)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not get start slot")
+		}
+		// We use the stategen replayer to fetch the finalized state and then
+		// replay it to the start slot of our checkpoint's epoch. The replayer
+		// only ever accesses our canonical history, so the state retrieved will
+		// always be the finalized state at that epoch.
+		s, err = p.ReplayerBuilder.ReplayerForSlot(startSlot).ReplayToSlot(ctx, startSlot)
 		if err != nil {
 			return nil, errors.Wrap(err, "could not get finalized state")
 		}
 	case "justified":
 		checkpoint := p.ChainInfoFetcher.CurrentJustifiedCheckpt()
-		s, err = p.StateGenService.StateByRoot(ctx, bytesutil.ToBytes32(checkpoint.Root))
+		startSlot, err := slots.EpochStart(checkpoint.Epoch)
```

Comment:
I think `startSlot` is a confusing name here. I know that `start` means the epoch start, but it is a little ambiguous because it is an argument for the replayer, where it represents the *end* slot the replayer should advance to. I'd suggest a name like `desiredSlot` or `target`.

Reply:
Alternatively, `epochStart` would be a little clearer. We know it's a slot from its type.